lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+                truncate: false,
+            },
+        }
+    }
+
+    pub fn read(&mut self, read: bool) -> &mut Self {
+        self.conf.read = read;
+        self
+    }
+
+    pub fn write(&mut self, write: bool) -> &mut Self {
+        self.conf.write = write;
+        self
+    }
+
+    pub fn append(&mut self, append: bool) -> &mut Self {
+        self.conf.append = append;
+        self
+    }
+
+    pub fn truncate(&mut self, truncate: bool) -> &mut Self {
+        self.conf.truncate = truncate;
+        self
+    }
+
+    pub fn create(&mut self, create: bool) -> &mut Self {
+        self.conf.create = create;
+        self
+    }
+
+    pub fn create_new(&mut self, create_new: bool) -> &mut Self {
+        self.conf.create_new = create_new;
+        self
+    }
+
+    pub fn open<P: AsRef<Path>>(&mut self, path: P) -> Result<Box<dyn VirtualFile>, FsError> {
+        self.opener.open(path.as_ref(), &self.conf)
+    }
+}
+
+/// This trait relies on your file closing when it goes out of scope via `Drop`
+#[typetag::serde(tag = "type")]
+pub trait VirtualFile: fmt::Debug + Send + Write + Read + Seek + 'static + Upcastable {
+    /// the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn last_accessed(&self) -> u64;
+
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64;
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+
+    /// set the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn set_last_accessed(&self, _last_accessed: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_accessed due to using the default implementation", self);
+    }
+
+    /// set the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn set_last_modified(&self, _last_modified: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_modified due to using the default implementation", self);
+    }
+
+    /// set the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn set_created_time(&self, _created_time: u64) {
+        debug!(
+            "{:?} did nothing in VirtualFile::set_created_time to using the default implementation",
+            self
+        );
+    }
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
+
+    /// Request deletion of the file
+    fn unlink(&mut self) -> Result<(), FsError>;
+
+    /// Store file contents and metadata to disk
+    /// Default implementation returns `Ok(())`.  You should implement this method if you care
+    /// about flushing your cache to permanent storage
+    fn sync_to_disk(&self) -> Result<(), FsError> {
+        Ok(())
+    }
```

Comment:
I think this only make sense for some kind of filesystems. Can we move it to other layer?

Reply:
Well it's fine for filesystems to not implement it, if this is in WasiFile then it's fine. But WASI needs a method like this to call for things to be emulated accurately

Reply:
Why not move it to the WASI filesystem layer?

Reply:
How would we? This is the layer through which WASI is interacting with the filesystem

Reply:
I think it's important to have not only these types of methods, but a lot more if we want this to be generally useful as a drop-in replacement abstraction for real file system usage...